### ISSUE TITLE
AnimatedVisualPlayer::PlayAsync should internally wait for the AnimatedVisual to load instead of silently failing

### ISSUE THREAD

**Issue by jaredhms** [microsoft/microsoft-ui-xaml#1484]
If PlayAsync is called before the AnimatedVisual is loaded, nothing happens, yet the IAsyncAction reports success.

The app developer has to write code like this, whereas it would make more sense for AnimatedVisualPlayer to handle this complexity internally.

`    

            // We have to wait until the player successfully loads its animation before we can call Play!
            auto notifyPropertyChanged = m_animatedVisualSource.as<winrt::Windows::UI::Xaml::Data::INotifyPropertyChanged>();
            m_animatedVisualPropertyChangedRegistration = notifyPropertyChanged.PropertyChanged(
                [player, loop = m_loop](const auto& sender, const winrt::Windows::UI::Xaml::Data::PropertyChangedEventArgs& args)
            {
                // If the visual is loaded, we're ready to PlayAsync!
                if (winrt::to_string(args.PropertyName()) == "IsImageLoadingCompleted")
                {
                    player.PlayAsync(0, 1.0, loop);
                }
            });

            // Kick off play now if the visuals are already loaded
            if (player.IsAnimatedVisualLoaded())
            {
                player.PlayAsync(0, 1.0, m_loop);
            }

`

An interesting additional bug is that calling PlayAsync before a visual is loaded also breaks AutoPlay (nothing plays after the visual finally loads).  Not sure if that should be filed as a separate issue.

**Comment by simeoncran** [comment]
This is by design. This mimics the behavior of most media (e.g. sound and movie) players.

A PlayOnNextLoadAsync could be added to give the behavior you want, however we figured that most scenarios that use asynchronously loaded content need finer control over what happens if the asynchronous loading doesn't complete on time or as expected. I.e. it's usually simpler to wait for loading, then play, then it is to have to cancel playing if the loading doesn't complete.

BTW, there is another way to react to the the AnimatedVisualPlayer getting content loaded, and that is to listen to the AnimatedVisualPlayer.IsAnimatedVisualLoaded dependency property. That is a more general way of handling async loading as it works for any asynchronous content, not just image loading.

**Comment by jaredhms** [comment]
There are enough knobs/toggles here to allow for more advanced usage if/when needed, but I don't understand why set source+play isn't the 'golden path' for this API, since that allows you to pick up the API and use it with little to no friction.  

More advanced developers can worry more about timing, but I'd expect that most developers won't need to worry about it. Our use case is tied into React Native: render a view, play animations, stop/cancel animations when destroying the view or transitioning state.  Requiring us to wire up dependency property changes to know when it's 'safe' to call Play is unnecessary burden.

**Comment by chingucoding** [comment]
I'm confused now. So is the behavior intentional? Should PlayAsync not wait for the animation? Or should we change that, so that calling PlayAsync waits for the animation to have finished loading, and then start playing?

If the latter is the case, I would be happy to try to fix this. :)

**Comment by ranjeshj** [comment]
@MikeHillberg @simeoncran 
Some options to reduce the code required for this common scenario.
* We could add another parameter to PlayAsync to optionally delay until animation has finished loading, but that would be a breaking change. 
* We could add a new overload like PlayOnNextLoadAsync 
* We could add a boolean DelayAsyncPlayUntilLoad which modifies the behavior of the current PlayAsync method.

**Comment by chingucoding** [comment]
Those are all great options. Probably adding an optional parameter is the easiest/most convenient way for developers and also keeps the API surface clean. Should we create a proposal for this?

**Comment by MikeHillberg** [comment]
I'd expect it to just work too, and for the more advanced case you could monitor events (is there an event)?

What do [BitmapImage.Play](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Media.Imaging.BitmapImage.Play) (aka AniGif) and [MediaPlayerElement](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.MediaPlayerElement) ([MediaPlayer.Play](https://docs.microsoft.com/uwp/api/Windows.Media.Playback.MediaPlayer.Play)) do?

**Comment by chingucoding** [comment]
Calling Play on the MediaPlayerElement before it has loaded successfully plays the animation.

The big problem however with that comparison is that all of those methods are not async but run sequentially, so setting source might only return upon successful loading.

**Comment by chingucoding** [comment]
@MikeHillberg @ranjeshj @StephenLPeters What are your thoughts on this? Should we change the API surface or should PlayAsync change here?

**Comment by simeoncran** [comment]
To make AVP start playing an asynchronously loaded animated visual as soon as it finishes loading:

        static void WaitForLoadThenPlayAsync(AnimatedVisualPlayer player)
        {
            if (player.IsAnimatedVisualLoaded)
            {
                _ = player.PlayAsync(0, 1, true);
            }
            else
            {
                player.RegisterPropertyChangedCallback(AnimatedVisualPlayer.IsAnimatedVisualLoadedProperty, (sender, dp) =>
                {
                    _ = player.PlayAsync(0, 1, true);
                });
            }
        }

This is similar to what was shown at the start of this issue, but it doesn't require referencing a particular animated visual source so you can just write it once. But this is overkill for your case because you don't care about responsiveness and you expect your images to load quickly, so...

_**Turn off asynchronous loading**_. This will allow you to use AVP as if you had a synchronously loaded animated visual, with behavior that should be the same as what you'd get from the web player, i.e. the Lottie will start playing immediately and the images will pop in as they load. Typically the images will appear almost immediately and the user will not be aware of the asynchrony. 

Lottiegen enables asynchronous loading of animated visuals that use images, but it generates a property so you can turn off asynchronous loading: IsAnimatedVisualSourceDynamic. For example, to auto-play a Lottie that has images without using any code behind:

        <muxc:AnimatedVisualPlayer x:Name="Player">
            <animatedVisuals:Splash_Intro_Large IsAnimatedVisualSourceDynamic="False" />
        </muxc:AnimatedVisualPlayer>

I suspect we should have made IsAnimatedVisualSourceDynamic default to false so as to avoid this issue bothering anyone. Tracking that here: https://github.com/windows-toolkit/Lottie-Windows/issues/340

**Comment by simeoncran** [comment]
@chingucoding right, I would advise against changing the design to make this rare case easier at the expense of making all cases harder to reason about. Is this actually causing anyone any pain that can't be solved by simply setting IsAnimatedVisualSourceDynamic to false?

**Comment by chingucoding** [comment]
I would argue that it's more in the range of "nice to have" rather than a major pain point, though I haven't worked that much with AVP. I can see why it would make sense for very simple scenarios to have PlayAsync to wait, however as you noted, more complex scenarios will get harder to reason about. I'm inclined to say that it might be better to leave as is then, @StephenLPeters @jaredhms, your thoughts?

**Comment by simeoncran** [comment]
What is the simple scenario where you'd prefer to have it wait rather than making the content non-asynchronous to begin with?

**Comment by chingucoding** [comment]
Thinking about it, when you can't make it non-asynchronous (e.g. loading files or something like that), you need to handle failure anyway. I can't think of a simple scenario where making the content non-asynchronous wouldn't work. Now I am even more convinced that the current is probably the best way to handle this, thank you @simeoncran for helping us/me this scenario better!

**Comment by chingucoding** [comment]
@StephenLPeters @ranjeshj Should this issue stay open or can this be closed given the recommendations from @simeoncran ?
